File: python_model/collect_data.py
#!/usr/bin/env python3
"""
BTC/USD 数据采集 - 同时从 Binance 和 Chainlink 获取数据。
Polymarket 使用 Chainlink 数据结算，必须使用相同数据源训练。

数据源:
- Binance: BTC/USDT 现货价格 (高频，用于特征)
- Chainlink: BTC/USD 预言机价格 (结算价格)
"""

import os
import time
from datetime import datetime, timedelta
from pathlib import Path
from typing import Optional

import pandas as pd
import requests
from web3 import Web3

# 超时配置
REQUEST_TIMEOUT = 30  # 网络请求超时（秒）
WEB3_TIMEOUT = 30  # Web3 调用超时（秒）
MAX_RETRIES = 3  # 最大重试次数
RETRY_DELAY = 1  # 重试延迟（秒）

DATA_DIR = Path(__file__).parent.parent / "data"
DATA_DIR.mkdir(exist_ok=True)

# === Chainlink 配置 ===
# Polygon Mainnet BTC/USD Price Feed
# 来源: https://docs.chain.link/data-feeds/price-feeds/addresses?network=polygon
CHAINLINK_BTC_USD_POLYGON = "0xc907E116054Ad103354f2D350FD2514433D57F6f"

# Ethereum Mainnet BTC/USD Price Feed (备用)
# 来源: https://docs.chain.link/data-feeds/price-feeds/addresses?network=ethereum
CHAINLINK_BTC_USD_ETH = "0xF4030086522a5bEEa4988F8cA5B36dbC97BeE88c"

# Chainlink AggregatorV3Interface ABI
CHAINLINK_ABI = [
    {
        "inputs": [],
        "name": "latestRoundData",
        "outputs": [
            {"name": "roundId", "type": "uint80"},
            {"name": "answer", "type": "int256"},
            {"name": "startedAt", "type": "uint256"},
            {"name": "updatedAt", "type": "uint256"},
            {"name": "answeredInRound", "type": "uint80"},
        ],
        "stateMutability": "view",
        "type": "function",
    },
    {
        "inputs": [],
        "name": "decimals",
        "outputs": [{"name": "", "type": "uint8"}],
        "stateMutability": "view",
        "type": "function",
    },
    {
        "inputs": [{"name": "_roundId", "type": "uint80"}],
        "name": "getRoundData",
        "outputs": [
            {"name": "roundId", "type": "uint80"},
            {"name": "answer", "type": "int256"},
            {"name": "startedAt", "type": "uint256"},
            {"name": "updatedAt", "type": "uint256"},
            {"name": "answeredInRound", "type": "uint80"},
        ],
        "stateMutability": "view",
        "type": "function",
    },
]

# RPC URLs (使用公共节点或配置自己的)
POLYGON_RPC = os.environ.get("POLYGON_RPC_URL", "https://polygon-rpc.com")
ETH_RPC = os.environ.get("ETH_RPC_URL", "https://eth.llamarpc.com")


def get_chainlink_price(
    rpc_url: str = POLYGON_RPC,
    contract_address: str = CHAINLINK_BTC_USD_POLYGON,
    timeout: int = WEB3_TIMEOUT,
) -> tuple[float, int]:
    """
    从 Chainlink 获取当前 BTC/USD 价格。
    
    Args:
        timeout: Web3 调用超时时间（秒）
    
    Returns:
        Tuple of (price, timestamp)
    """
    w3 = Web3(Web3.HTTPProvider(rpc_url, request_kwargs={"timeout": timeout}))
    contract = w3.eth.contract(
        address=Web3.to_checksum_address(contract_address),
        abi=CHAINLINK_ABI,
    )
    
    try:
        _, answer, _, updated_at, _ = contract.functions.latestRoundData().call()
        decimals = contract.functions.decimals().call()
    except Exception as e:
        raise Exception(f"Chainlink 调用超时或失败 (超时设置: {timeout}秒): {e}")
    
    price = answer / (10 ** decimals)
    return price, updated_at


def fetch_chainlink_historical(
    rpc_url: str = POLYGON_RPC,
    contract_address: str = CHAINLINK_BTC_USD_POLYGON,
    num_rounds: int = 1000,
    timeout: int = WEB3_TIMEOUT,
    max_consecutive_failures: int = 10,
) -> pd.DataFrame:
    """
    从 Chainlink 获取历史价格数据 (通过遍历历史 round)。
    
    注意: Chainlink 更新频率约每小时或价格变动 >0.5% 时更新。
    
    Args:
        timeout: Web3 调用超时时间（秒）
        max_consecutive_failures: 最大连续失败次数
    
    Returns:
        DataFrame with timestamp and price
    """
    print(f"[Chainlink] 获取历史数据, contract: {contract_address}")
    print(f"  超时设置: {timeout}秒")
    
    w3 = Web3(Web3.HTTPProvider(rpc_url, request_kwargs={"timeout": timeout}))
    contract = w3.eth.contract(
        address=Web3.to_checksum_address(contract_address),
        abi=CHAINLINK_ABI,
    )
    
    # 获取当前 round
    try:
        round_id, answer, _, updated_at, _ = contract.functions.latestRoundData().call()
        decimals = contract.functions.decimals().call()
    except Exception as e:
        raise Exception(f"Chainlink 调用超时或失败 (超时设置: {timeout}秒): {e}")
    
    data = []
    current_round = round_id
    consecutive_failures = 0
    
    for i in range(num_rounds):
        try:
            rid, ans, started, updated, _ = contract.functions.getRoundData(current_round).call()
            price = ans / (10 ** decimals)
            
            data.append({
                "timestamp": datetime.utcfromtimestamp(updated),
                "price": price,
                "round_id": rid,
            })
            
            consecutive_failures = 0  # 重置失败计数
            
            # 前一个 round (Chainlink round ID 结构复杂，简单递减可能不工作)
            # 使用 phase + aggregator round 结构
            current_round -= 1
            
            if i % 100 == 0:
                print(f"  已获取 {i+1}/{num_rounds} rounds")
                
        except Exception as e:
            consecutive_failures += 1
            if consecutive_failures >= max_consecutive_failures:
                print(f"  连续失败 {consecutive_failures} 次，停止获取")
                break
            if i % 10 == 0:  # 减少错误输出频率
                print(f"  Round {current_round} 获取失败 ({consecutive_failures}/{max_consecutive_failures}): {e}")
            current_round -= 1
            continue
        
        time.sleep(0.1)  # Rate limiting
    
    df = pd.DataFrame(data)
    df = df.sort_values("timestamp")
    df = df.set_index("timestamp")
    
    return df


def fetch_binance_klines(
    symbol: str = "BTCUSDT",
    interval: str = "1m",
    start_time: Optional[datetime] = None,
    end_time: Optional[datetime] = None,
    limit: int = 1000,
    timeout: int = REQUEST_TIMEOUT,
    max_retries: int = MAX_RETRIES,
) -> pd.DataFrame:
    """
    从 Binance 获取 K 线数据。
    
    Args:
        timeout: 请求超时时间（秒）
        max_retries: 最大重试次数
    """
    url = "https://api.binance.com/api/v3/klines"
    
    params = {
        "symbol": symbol,
        "interval": interval,
        "limit": limit,
    }
    
    if start_time:
        params["startTime"] = int(start_time.timestamp() * 1000)
    if end_time:
        params["endTime"] = int(end_time.timestamp() * 1000)
    
    # 重试机制
    last_error = None
    for attempt in range(max_retries):
        try:
            response = requests.get(url, params=params, timeout=timeout)
            response.raise_for_status()
            klines = response.json()
            break
        except requests.exceptions.Timeout:
            last_error = f"请求超时（{timeout}秒）"
            if attempt < max_retries - 1:
                wait_time = RETRY_DELAY * (attempt + 1)
                print(f"  重试 {attempt + 1}/{max_retries}... ({wait_time}秒后)")
                time.sleep(wait_time)
            else:
                raise requests.exceptions.Timeout(f"{last_error}，已重试 {max_retries} 次")
        except requests.exceptions.RequestException as e:
            last_error = str(e)
            if attempt < max_retries - 1:
                wait_time = RETRY_DELAY * (attempt + 1)
                print(f"  请求失败，重试 {attempt + 1}/{max_retries}... ({wait_time}秒后)")
                time.sleep(wait_time)
            else:
                raise requests.exceptions.RequestException(f"请求失败: {last_error}，已重试 {max_retries} 次")
    else:
        raise Exception(f"获取数据失败: {last_error}")
    
    df = pd.DataFrame(
        klines,
        columns=[
            "open_time", "open", "high", "low", "close", "volume",
            "close_time", "quote_volume", "trades",
            "taker_buy_base", "taker_buy_quote", "ignore",
        ],
    )
    
    df["timestamp"] = pd.to_datetime(df["open_time"], unit="ms", utc=True)
    for col in ["open", "high", "low", "close", "volume", "quote_volume"]:
        df[col] = df[col].astype(float)
    df["trades"] = df["trades"].astype(int)
    
    df = df[["timestamp", "open", "high", "low", "close", "volume", "quote_volume", "trades"]]
    df = df.set_index("timestamp")
    
    return df


def fetch_binance_historical(
    days: int = 90,
    interval: str = "1m",
    max_failures: int = 5,  # 最大连续失败次数
) -> pd.DataFrame:
    """
    批量获取 Binance 历史数据。
    
    Args:
        days: 获取天数
        interval: K线间隔
        max_failures: 最大连续失败次数，超过后停止
    """
    print(f"[Binance] 获取 {days} 天 {interval} 数据...")
    print(f"  超时设置: {REQUEST_TIMEOUT}秒，最大重试: {MAX_RETRIES}次")
    
    all_data = []
    end_time = datetime.now(tz=None)  # naive datetime
    start_time = end_time - timedelta(days=days)
    
    current_start = start_time
    consecutive_failures = 0
    batch_count = 0
    
    while current_start < end_time:
        batch_count += 1
        try:
            df = fetch_binance_klines(
                start_time=current_start,
                end_time=end_time,
            )
            if len(df) == 0:
                print(f"  批次 {batch_count}: 未获取到数据，可能已到达最新时间")
                break
            
            all_data.append(df)
            consecutive_failures = 0  # 重置失败计数
            
            # 转换为 naive datetime 进行比较
            last_time = df.index.max()
            if hasattr(last_time, 'tz_localize'):
                last_time = last_time.tz_localize(None)
            current_start = last_time.to_pydatetime() + timedelta(minutes=1)
            
            progress = (current_start - start_time).total_seconds() / (end_time - start_time).total_seconds()
            print(f"  批次 {batch_count}: {df.index.min()} - {df.index.max()}, 共 {len(df)} 条 (进度: {progress:.1%})")
            
        except Exception as e:
            consecutive_failures += 1
            print(f"  批次 {batch_count} 获取失败 ({consecutive_failures}/{max_failures}): {e}")
            
            if consecutive_failures >= max_failures:
                print(f"  ⚠️  连续失败 {consecutive_failures} 次，停止获取")
                break
            
            # 失败后等待更长时间
            time.sleep(RETRY_DELAY * consecutive_failures)
            continue
        
        time.sleep(0.5)  # 正常请求间的延迟
    
    if not all_data:
        raise ValueError(f"未获取到任何数据。请检查网络连接或 Binance API 是否可用（超时设置: {REQUEST_TIMEOUT}秒）")
    
    combined = pd.concat(all_data)
    combined = combined[~combined.index.duplicated(keep="last")]
    combined = combined.sort_index()
    
    print(f"✅ 完成！共获取 {len(combined)} 条数据")
    return combined


def resample_to_15m(df: pd.DataFrame) -> pd.DataFrame:
    """
    将 1 分钟数据重采样为 15 分钟。
    
    关键: 对齐到固定的 15 分钟窗口边界 (00:00, 00:15, 00:30, 00:45)
    这与 Polymarket 的结算窗口一致。
    """
    # 确保索引是 datetime
    if not isinstance(df.index, pd.DatetimeIndex):
        df.index = pd.to_datetime(df.index, utc=True)
    
    # 重采样到 15 分钟，对齐到窗口开始
    # origin='start_day' 确保从 00:00 开始对齐
    df_15m = df.resample("15min", origin="start_day").agg({
        "open": "first",
        "high": "max",
        "low": "min",
        "close": "last",
        "volume": "sum",
        "quote_volume": "sum",
        "trades": "sum",
    }).dropna()
    
    return df_15m


def load_existing_data(filename: str) -> Optional[pd.DataFrame]:
    """加载已存在的数据文件。"""
    filepath = DATA_DIR / filename
    if not filepath.exists():
        return None
    
    df = pd.read_csv(filepath, parse_dates=["timestamp"], index_col="timestamp")
    df.index = pd.to_datetime(df.index, utc=True)
    return df


def save_data(df: pd.DataFrame, filename: str) -> None:
    """保存数据到 CSV。"""
    filepath = DATA_DIR / filename
    df.to_csv(filepath)
    print(f"已保存 {len(df)} 条数据到 {filepath}")


def collect_data(days: int = 90) -> tuple[pd.DataFrame, pd.DataFrame]:
    """
    收集数据的主函数。
    
    Returns:
        Tuple of (1分钟数据, 15分钟数据)
    """
    # 1. 检查已存在的数据
    existing_1m = load_existing_data("btc_1m.csv")
    
    if existing_1m is not None and len(existing_1m) > 0:
        last_time = existing_1m.index.max()
        time_since_last = datetime.now(tz=existing_1m.index.tzinfo) - last_time
        
        if time_since_last < timedelta(minutes=5):
            print("数据已是最新")
            df_15m = resample_to_15m(existing_1m)
            save_data(df_15m, "btc_15m.csv")
            return existing_1m, df_15m
        
        # 增量更新
        print(f"增量更新，从 {last_time} 开始...")
        new_df = fetch_binance_klines(start_time=last_time.to_pydatetime())
        
        combined = pd.concat([existing_1m, new_df])
        combined = combined[~combined.index.duplicated(keep="last")]
        combined = combined.sort_index()
        df_1m = combined
    else:
        # 全量获取
        df_1m = fetch_binance_historical(days=days)
    
    # 2. 保存 1 分钟数据
    save_data(df_1m, "btc_1m.csv")
    
    # 3. 重采样为 15 分钟
    df_15m = resample_to_15m(df_1m)
    save_data(df_15m, "btc_15m.csv")
    
    # 4. 尝试获取 Chainlink 数据 (可选，用于验证)
    try:
        cl_price, cl_time = get_chainlink_price()
        print(f"\n[Chainlink] 当前价格: ${cl_price:,.2f} (更新于 {datetime.utcfromtimestamp(cl_time)})")
        
        binance_price = df_1m["close"].iloc[-1]
        diff = abs(cl_price - binance_price)
        diff_pct = diff / binance_price * 100
        print(f"[对比] Binance: ${binance_price:,.2f}, 差异: ${diff:.2f} ({diff_pct:.3f}%)")
    except Exception as e:
        print(f"[Chainlink] 获取失败 (需要配置 RPC): {e}")
    
    return df_1m, df_15m


def load_data(interval: str = "15m") -> pd.DataFrame:
    """
    加载数据。
    
    Args:
        interval: "1m" 或 "15m"
    """
    filename = f"btc_{interval}.csv"
    filepath = DATA_DIR / filename
    
    if not filepath.exists():
        raise FileNotFoundError(f"数据文件不存在: {filepath}, 请先运行 collect_data()")
    
    df = pd.read_csv(filepath, parse_dates=["timestamp"], index_col="timestamp")
    df.index = pd.to_datetime(df.index, utc=True)
    return df


if __name__ == "__main__":
    df_1m, df_15m = collect_data(days=90)
    
    print(f"\n=== 数据摘要 ===")
    print(f"1 分钟数据: {len(df_1m)} 条, {df_1m.index.min()} ~ {df_1m.index.max()}")
    print(f"15 分钟数据: {len(df_15m)} 条, {df_15m.index.min()} ~ {df_15m.index.max()}")
    print(f"\n15 分钟数据样本:\n{df_15m.tail()}")

File: node_bot/src/collectData.ts
#!/usr/bin/env node
/**
 * BTC/USD 数据采集 - 从 Binance 获取历史数据
 * 
 * 功能：
 * - 从 Binance 获取 1 分钟 K 线数据
 * - 重采样为 15 分钟数据
 * - 保存为 CSV 文件（与 Python 版本兼容）
 * - 完整的超时和错误处理，不会卡住
 */

import axios, { AxiosError } from "axios";
import { writeFileSync, mkdirSync, existsSync } from "fs";
import { resolve, dirname } from "path";
import { fileURLToPath } from "url";

const __filename = fileURLToPath(import.meta.url);
const __dirname = dirname(__filename);

// 配置
const REQUEST_TIMEOUT = 30000; // 30秒超时
const MAX_RETRIES = 3; // 最大重试次数
const RETRY_DELAY = 1000; // 重试延迟（毫秒）
const MAX_FAILURES = 5; // 最大连续失败次数
const BATCH_DELAY = 500; // 批次间延迟（毫秒）
const BINANCE_API = "https://api.binance.com/api/v3/klines";

// 数据目录
const DATA_DIR = resolve(__dirname, "../../data");
const CSV_1M = resolve(DATA_DIR, "btc_1m.csv");
const CSV_15M = resolve(DATA_DIR, "btc_15m.csv");

// 确保数据目录存在
mkdirSync(DATA_DIR, { recursive: true });

// Binance API 返回的是数组格式，每个 K 线是一个数组：
// [open_time, open, high, low, close, volume, close_time, quote_volume, trades, taker_buy_base, taker_buy_quote, ignore]
type BinanceKline = [
  number, // open_time
  string, // open
  string, // high
  string, // low
  string, // close
  string, // volume
  number, // close_time
  string, // quote_volume
  number, // trades
  string, // taker_buy_base
  string, // taker_buy_quote
  string  // ignore
];

interface Candle {
  timestamp: Date;
  open: number;
  high: number;
  low: number;
  close: number;
  volume: number;
  quote_volume: number;
  trades: number;
}

/**
 * 从 Binance 获取 K 线数据（带超时和重试）
 */
async function fetchBinanceKlines(
  symbol: string = "BTCUSDT",
  interval: string = "1m",
  startTime?: number,
  endTime?: number,
  limit: number = 1000
): Promise<Candle[]> {
  const params: Record<string, string> = {
    symbol,
    interval,
    limit: limit.toString(),
  };

  if (startTime) {
    params.startTime = startTime.toString();
  }
  if (endTime) {
    params.endTime = endTime.toString();
  }

  let lastError: Error | null = null;

  for (let attempt = 0; attempt < MAX_RETRIES; attempt++) {
    try {
      const response = await axios.get<unknown>(BINANCE_API, {
        params,
        timeout: REQUEST_TIMEOUT,
        validateStatus: (status) => status === 200,
      });

      // 检查响应数据
      if (!response.data || !Array.isArray(response.data)) {
        throw new Error(`API 返回格式错误: ${JSON.stringify(response.data).substring(0, 200)}`);
      }

      if (response.data.length === 0) {
        return [];
      }

      const candles: Candle[] = (response.data as BinanceKline[]).map((k) => {
        // Binance 返回的格式是数组：[open_time, open, high, low, close, volume, close_time, quote_volume, trades, ...]
        const timestamp = new Date(k[0]); // open_time 是毫秒时间戳
        if (isNaN(timestamp.getTime())) {
          throw new Error(`无效的时间戳: ${k[0]}`);
        }
        return {
          timestamp,
          open: parseFloat(k[1]),      // open
          high: parseFloat(k[2]),      // high
          low: parseFloat(k[3]),       // low
          close: parseFloat(k[4]),     // close
          volume: parseFloat(k[5]),    // volume
          quote_volume: parseFloat(k[7]), // quote_volume
          trades: k[8],                // trades
        };
      });

      return candles;
    } catch (error) {
      if (axios.isAxiosError(error)) {
        const axiosError = error as AxiosError;
        if (axiosError.code === "ECONNABORTED" || axiosError.message.includes("timeout")) {
          lastError = new Error(`请求超时（${REQUEST_TIMEOUT}ms）`);
        } else {
          lastError = new Error(`请求失败: ${axiosError.message}`);
        }
      } else {
        lastError = error as Error;
      }

      if (attempt < MAX_RETRIES - 1) {
        const waitTime = RETRY_DELAY * (attempt + 1);
        console.log(`  重试 ${attempt + 1}/${MAX_RETRIES}... (${waitTime}ms后)`);
        await new Promise((resolve) => setTimeout(resolve, waitTime));
      } else {
        throw new Error(`${lastError.message}，已重试 ${MAX_RETRIES} 次`);
      }
    }
  }

  throw new Error(`获取数据失败: ${lastError?.message ?? "未知错误"}`);
}

/**
 * 批量获取历史数据
 */
async function fetchBinanceHistorical(
  days: number = 90,
  interval: string = "1m"
): Promise<Candle[]> {
  console.log(`[Binance] 获取 ${days} 天 ${interval} 数据...`);
  console.log(`  超时设置: ${REQUEST_TIMEOUT}ms，最大重试: ${MAX_RETRIES}次`);

  const allData: Candle[] = [];
  const endTime = new Date();
  const startTime = new Date(endTime.getTime() - days * 24 * 60 * 60 * 1000);

  let currentStart = startTime;
  let consecutiveFailures = 0;
  let batchCount = 0;

  while (currentStart < endTime) {
    batchCount++;
    try {
      const candles = await fetchBinanceKlines(
        "BTCUSDT",
        interval,
        currentStart.getTime(),
        endTime.getTime()
      );

      if (candles.length === 0) {
        console.log(`  批次 ${batchCount}: 未获取到数据，可能已到达最新时间`);
        break;
      }

      allData.push(...candles);
      consecutiveFailures = 0; // 重置失败计数

      // 更新起始时间为最后一根 K 线的下一个分钟
      const lastCandle = candles[candles.length - 1];
      if (lastCandle) {
        currentStart = new Date(lastCandle.timestamp.getTime() + 60 * 1000);
      } else {
        break;
      }

      const progress =
        ((currentStart.getTime() - startTime.getTime()) /
          (endTime.getTime() - startTime.getTime())) *
        100;
      const firstTime = candles[0]?.timestamp.toISOString() ?? "";
      const lastTime = lastCandle?.timestamp.toISOString() ?? "";
      console.log(
        `  批次 ${batchCount}: ${firstTime} - ${lastTime}, 共 ${candles.length} 条 (进度: ${progress.toFixed(1)}%)`
      );
    } catch (error) {
      consecutiveFailures++;
      const errorMsg = error instanceof Error ? error.message : String(error);
      console.log(
        `  批次 ${batchCount} 获取失败 (${consecutiveFailures}/${MAX_FAILURES}): ${errorMsg}`
      );

      if (consecutiveFailures >= MAX_FAILURES) {
        console.log(`  ⚠️  连续失败 ${consecutiveFailures} 次，停止获取`);
        break;
      }

      // 失败后等待更长时间
      await new Promise((resolve) =>
        setTimeout(resolve, RETRY_DELAY * consecutiveFailures)
      );
      continue;
    }

    // 正常请求间的延迟
    await new Promise((resolve) => setTimeout(resolve, BATCH_DELAY));
  }

  if (allData.length === 0) {
    throw new Error(
      `未获取到任何数据。请检查网络连接或 Binance API 是否可用（超时设置: ${REQUEST_TIMEOUT}ms）`
    );
  }

  // 去重并排序
  const uniqueMap = new Map<number, Candle>();
  for (const candle of allData) {
    const key = candle.timestamp.getTime();
    if (!uniqueMap.has(key) || uniqueMap.get(key)!.timestamp < candle.timestamp) {
      uniqueMap.set(key, candle);
    }
  }

  const sorted = Array.from(uniqueMap.values()).sort(
    (a, b) => a.timestamp.getTime() - b.timestamp.getTime()
  );

  console.log(`✅ 完成！共获取 ${sorted.length} 条数据`);
  return sorted;
}

/**
 * 将 1 分钟数据重采样为 15 分钟数据
 */
function resampleTo15m(candles: Candle[]): Candle[] {
  if (candles.length === 0) {
    return [];
  }

  // 按 15 分钟窗口分组
  const windows = new Map<string, Candle[]>();

  for (const candle of candles) {
    // 对齐到 15 分钟窗口开始（00:00, 00:15, 00:30, 00:45）
    const timestamp = candle.timestamp;
    const minutes = timestamp.getUTCMinutes();
    const alignedMinutes = Math.floor(minutes / 15) * 15;
    const windowStart = new Date(timestamp);
    windowStart.setUTCMinutes(alignedMinutes, 0, 0);

    const key = windowStart.toISOString();
    if (!windows.has(key)) {
      windows.set(key, []);
    }
    windows.get(key)!.push(candle);
  }

  // 聚合每个窗口
  const result: Candle[] = [];

  for (const [windowStartStr, windowCandles] of windows.entries()) {
    if (windowCandles.length === 0) continue;

    const sorted = windowCandles.sort(
      (a, b) => a.timestamp.getTime() - b.timestamp.getTime()
    );

    const open = sorted[0]!.open;
    const close = sorted[sorted.length - 1]!.close;
    const high = Math.max(...sorted.map((c) => c.high));
    const low = Math.min(...sorted.map((c) => c.low));
    const volume = sorted.reduce((sum, c) => sum + c.volume, 0);
    const quoteVolume = sorted.reduce((sum, c) => sum + c.quote_volume, 0);
    const trades = sorted.reduce((sum, c) => sum + c.trades, 0);

    result.push({
      timestamp: new Date(windowStartStr),
      open,
      high,
      low,
      close,
      volume,
      quote_volume: quoteVolume,
      trades,
    });
  }

  return result.sort((a, b) => a.timestamp.getTime() - b.timestamp.getTime());
}

/**
 * 保存数据为 CSV（与 Python 版本兼容）
 */
function saveData(candles: Candle[], filepath: string): void {
  if (candles.length === 0) {
    throw new Error("没有数据可保存");
  }

  // CSV 头部
  const headers = [
    "timestamp",
    "open",
    "high",
    "low",
    "close",
    "volume",
    "quote_volume",
    "trades",
  ];

  // 生成 CSV 内容
  const lines: string[] = [headers.join(",")];

  for (const candle of candles) {
    const row = [
      candle.timestamp.toISOString(),
      candle.open.toString(),
      candle.high.toString(),
      candle.low.toString(),
      candle.close.toString(),
      candle.volume.toString(),
      candle.quote_volume.toString(),
      candle.trades.toString(),
    ];
    lines.push(row.join(","));
  }

  writeFileSync(filepath, lines.join("\n"), "utf-8");
  console.log(`已保存 ${candles.length} 条数据到 ${filepath}`);
}

/**
 * 主函数
 */
async function main(): Promise<void> {
  const days = process.argv[2] ? parseInt(process.argv[2], 10) : 7;

  if (isNaN(days) || days < 1) {
    console.error("错误: days 参数必须是正整数");
    process.exit(1);
  }

  try {
    console.log("=".repeat(60));
    console.log("BTC 数据采集");
    console.log("=".repeat(60));

    // 1. 获取 1 分钟数据
    console.log("\n[1/3] 获取 1 分钟数据...");
    const df1m = await fetchBinanceHistorical(days, "1m");
    saveData(df1m, CSV_1M);

    // 2. 重采样为 15 分钟
    console.log("\n[2/3] 重采样为 15 分钟...");
    const df15m = resampleTo15m(df1m);
    saveData(df15m, CSV_15M);

    // 3. 输出摘要
    console.log("\n[3/3] 数据摘要:");
    console.log(`1 分钟数据: ${df1m.length} 条`);
    if (df1m.length > 0) {
      console.log(
        `  时间范围: ${df1m[0]!.timestamp.toISOString()} ~ ${
          df1m[df1m.length - 1]!.timestamp.toISOString()
        }`
      );
    }
    console.log(`15 分钟数据: ${df15m.length} 条`);
    if (df15m.length > 0) {
      console.log(
        `  时间范围: ${df15m[0]!.timestamp.toISOString()} ~ ${
          df15m[df15m.length - 1]!.timestamp.toISOString()
        }`
      );
    }

    console.log("\n✅ 完成!");
  } catch (error) {
    console.error("\n❌ 错误:", error instanceof Error ? error.message : String(error));
    process.exit(1);
  }
}

// 如果直接运行此脚本（通过 tsx 或 node）
// 使用更简单可靠的方法：检查是否直接运行
if (import.meta.url.endsWith(process.argv[1]?.replace(/\\/g, '/') || '') || 
    process.argv[1]?.includes('collectData')) {
  main().catch((err) => {
    console.error('执行失败:', err);
    process.exit(1);
  });
}

// 导出函数供其他模块使用
export { fetchBinanceHistorical, resampleTo15m, saveData, fetchBinanceKlines };
export type { Candle };

